Openvpn experiment spec

ainghazal opened the pull request:
## Checklist

- [x] I have read the [contribution guidelines](https://github.com/ooni/spec/blob/master/CONTRIBUTING.md)
- [x] reference issue for this pull request: https://github.com/ooni/probe/issues/2202
- [x] related ooni/probe-cli pull request: https://github.com/ooni/probe-cli/pull/845
- [x] If I changed a spec, I also bumped its version number and/or date

Location of the issue tracker: https://github.com/ooni/probe

## Description

This is a first draft of the openvpn experiment spec, in its current iteration.

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+    "minivpn_version": "(devel)",
+    "obfs4_version": "",
+    "success": true
+    "stages": [
```
Would it perhaps make sense to use the `network_events` data format (https://github.com/ooni/spec/blob/master/data-formats/df-008-netevents.md) for these sorts of events?

It seems like it fits pretty well within that model. You could use `transaction_id` in place of `op_id`.

ainghazal replied:
It makes sense - @bassosimone had previously recommended to mimic more closely that model. as per the df-008 operations are more restricted, but I guess I can specify that. Looking at, for instance, ts-029-signal I think it also makes sense to include a `failed_operation` field (int) so that we can easily track what's the failed state.

ainghazal replied:
will change the key `stages` to `network_events` too.

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+      },
+      "t0": 0.000239501,
+      "t": 0.045310671,
+      "transaction_id": 1
```
What is the relationship between this `transaction_id` and other parts of the document?

ainghazal replied:
I think this is always to be 1, since it's getting the first TCPConnect (*model.ArchivalTCPConnectResult):

`trace.FirstTCPConnectOrNil()`

https://github.com/ooni/probe-cli/blob/8dd1ae28a3593fcb7d083c0f3d9fb1263e648159/internal/measurexlite/dialer.go#L133

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+    },
+    "failure": null,
+    "error": null,
+    "pings": [
```
I'm not sure it's a problem, but we might want to think about the fact that we currently have 3 different structures for the key `pings` (see: https://github.com/ooni/spec/blob/master/nettests/ts-033-tlsping.md & https://github.com/ooni/spec/blob/master/nettests/ts-032-tcpping.md & https://github.com/ooni/spec/blob/master/nettests/ts-031-quicping.md).

For example we might care to disambiguate all these different pings by using a clearly distinct key (ex. `icmp_pings`) so that somebody parsing the data isn't lead into thinking they are the same thing.

ainghazal replied:
makes sense to rename as `icmp_pings`. it's possible that we end up internalizing the helper library too (it lives under github.com/ooni/minivpn/extras right now, but it needs some refactor to accomodate the differences between openvpn and wg tunnels).

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+        },
+        "response": {
+          "code": 200,
+          "headers": {
```
Is `headers_list` ommitted here for brevity, or does it in fact not generate `headers_list`, but only the `headers` map?

I checked the output sample and could not find `headers_list` there either.

hellais replied:
We discussed that it might be that the older measurement library is being used here and that's why some fields are different/missing.

TODO: check with @bassosimone to know if that is the case.

ainghazal replied:
@bassosimone confirmed I should be using `internal/model.ArchivalXXX` structs instead, since `measurex` is being deprecated. I'll update the PR.

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+          "x_body_is_utf8": true
+        },
+        "t": 0.558008438,
+        "started": 8.3589e-05,
```
Shouldn't this be `t0`?

ainghazal replied:
see comment above; whatever logic is in use is coming directly from the archivalHTTP_ implementation. For what I see, t0 is basically `requests[0].started`

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+- `obfs4_version` (`string`) contains the version of the `obfs4` library
+  that is used in the ooni probe build used for the experiment,
+- `failure` (`string`; nullable) conforms to `df-007-errors`,
+- `error` (`string`; nullable) raw error message, conforms to `df-007-errors`,
```
What's the difference between `failure` and `error`?

ainghazal replied:
I think I was initially following the vanilla tor spec, but now I see error was there for backward compatibility reasons. I will remove in favor of failure.

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+        - `ttls`: (`int`): raw ttl value,
+    - `pkt_rcv` (`int`): how many packets were received,
+    - `pkt_snt` (`int`): how many packets were sent,
+    - `min_rtt` (`float`): the min value for all the pings,
```
```suggestion
    - `min_rtt` (`float`): the minimum rtt value for all the pings,
```

ainghazal replied:
fixed

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+    - `pkt_rcv` (`int`): how many packets were received,
+    - `pkt_snt` (`int`): how many packets were sent,
+    - `min_rtt` (`float`): the min value for all the pings,
+    - `max_rtt` (`float`): the max value for all the pings,
```
```suggestion
    - `max_rtt` (`float`): the maximum rtt value for all the pings,
```

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+    - `pkt_snt` (`int`): how many packets were sent,
+    - `min_rtt` (`float`): the min value for all the pings,
+    - `max_rtt` (`float`): the max value for all the pings,
+    - `avg_rtt` (`float`): the avg value for all the pings,
```
```suggestion
    - `avg_rtt` (`float`): the average of the rtt for all the pings towards this target,
```

ainghazal replied:
fixed

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+    - `min_rtt` (`float`): the min value for all the pings,
+    - `max_rtt` (`float`): the max value for all the pings,
+    - `avg_rtt` (`float`): the avg value for all the pings,
+    - `std_rtt` (`float`): the std value for all the pings,
```
```suggestion
    - `std_rtt` (`float`): the standard deviation of the rtt for all the pings towards this target,
```

ainghazal replied:
fixed

hellais commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,710 @@
+    - `max_rtt` (`float`): the max value for all the pings,
+    - `avg_rtt` (`float`): the avg value for all the pings,
+    - `std_rtt` (`float`): the std value for all the pings,
+    - `error` (`string`; nullable): any error during the ping operation. Conforms to `df-007-errors`.
```
For consistency with other data formats, this should maybe be called `failure` (unless it's something different, see: https://github.com/ooni/spec/pull/255/files#r1018916989)?

ainghazal commented:
Pushed a new commit addressing comments. I still have to ask @bassosimone for possible replacement of the several measurement archivals in use (I might be using outdated operation wrappers).

ainghazal commented:
bumped to version 0.0.16; added ability to fetch arbitrary URLs.

bassosimone commented on data-formats/df-001-httpt.md:
```diff
@@ -122,6 +122,10 @@ are multiple values for the same header key. See below the definition of
 - `tor` (`TorInfo`): this is an object containing information on the
 instance of `tor` that we may be using for measuring.
 
+- `vpn` (`VPNInfo`): similarly to `tor`, this is an object containing
```
```suggestion
- `vpn` (`VPNInfo`; optional): similarly to `tor`, this is an object containing
```

We dont' want to include this field for every measurement like we (sadly) do for `tor`. We can also probably make `tor` optional without creating too much damage, but that would be out of scope here.

ainghazal replied:
added

bassosimone commented on data-formats/df-001-httpt.md:
```diff
@@ -222,6 +226,25 @@ if we're not using tor or this information is not available.
+  gateway belongs to, or `null` if we're not using a VPN or this information is
+  not available.
+
+- `is_vpn` (`bool`): true if we're using a VPN, false otherwise.
```
If we're making the whole structure optional, we can avoid making its internal fields nullable. Likewise, we can avoid having a field called `is_vpn` that would always be `true`. (Historically, the tor structure is a such because the precursor of web connectivity was sending the same request over tor and cleartext, hence the data format ended up having a `tor` structure that we carry along in any measurement.)

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+# Specification version number
+
+2022-10-20
```
We need to update this to the date of $today

ainghazal replied:
done

bassosimone commented on nettests/ts-039-openvpn.md:
````diff
@@ -0,0 +1,688 @@
+
+```
+vpn://{provider_name}.{protocol}/?transport=tcp&obfuscation=obfs4&address={ip}:{port}
+```
````
Please add an unordered list here where you explain the semantics of each parameter, like we normally do when you describe data structures.

ainghazal replied:
done

bassosimone commented on nettests/ts-039-openvpn.md:
````diff
@@ -0,0 +1,688 @@
+"SafeCert": "base64:deadbeef",
+"SafeKey": "base64:deadbeef",
+```
+
````
```suggestion
```

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+options.
+
+Altogether, they constitute a basic subset of the valid configuration
+parameters for the reference openvpn implementation (`man 8 openvpn`).
```
I think this section would be more readable if you would create subsections for URL and other options.

bassosimone commented on nettests/ts-039-openvpn.md:
````diff
@@ -0,0 +1,688 @@
+```
+"SafeCa": "base64:deadbeef",
+"SafeCert": "base64:deadbeef",
+"SafeKey": "base64:deadbeef",
````
I think it would be beneficial to explicit where these options would be set. Perhaps this could be helped by showing a subset of the toplevel structure including only the options and by referring to the base spec.

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+Other valid options, that will be passed to the openvpn invocation, are:
+
+- `cipher` (one of: `AES-256-GCM`, `AES-128-GCM`, `AES-256-CBC`, `AES-128-CBC`).
+- `auth` (one of: `SHA1`, `SHA512`).
```
I think you should explain with one sentence each of the possible options. Also beware that, IIUC, we need to use camel case names for options here.

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+Additionally, the user can specify any amount of comma-separated urls that
+wants to be fetched through the tunnel. 
+
+- `URLs` (`string`, optional).
```
This is where I am sad that the data format only allows a list of strings. However, I the minioonirun v2 spec we allow the options to be a map from string to any. I think something is wrong in the base data format used for options and we should instead allow a map from string to any.

ainghazal replied:
yea it seemed weird. I blindly followed cargo-cult while doing this. a new revision of the base format should be possible, isn't it?

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+# Test description
+
+There are, at the moment, three distinct stages in the OpenVPN experiments:
+`tunnel initialization`, `icmp-ping`, and `urlgrabber`.
```
Is there a reason why you're using a space separator and a dash separator inconsistently?

ainghazal replied:
probably some weird bias towards what a sintagm and a modificator are, but you're right. removing the dash.

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+
+An OpenVPN connection tries, sequentially, the following steps:
+
+1. Initializes cryptographic material (local only, not expected to fail).
```
Can you add an empty line between each point thus facilitating vi-like `{` and `}` navigation?

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+   error, including a timeout if a reply is not received after a reasonable
+   amount of time).
+5. Data channel initialization: client sends a control packet (expects a server
+  ACK), and a request for the server to push options. Upon receiving options
+  from the server, the data channel is initialized and ready to use.
```
```suggestion
   ACK), and a request for the server to push options. Upon receiving options
   from the server, the data channel is initialized and ready to use.
```

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+As reference targets, we measure one cloud provider, the tunnel gateway itself,
+and a box in a known geolocation.
+
+## 3. urlgrabber
```
```suggestion
## 3. URLGrabber
```

Suggesting because above you have ICMP-Pings

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+## Parent data format
+
+We include data from the following parent formats:
+
```
```suggestion
```

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+    "bootstrap_time": 1.5671859910000001,
+    "network_events": [
+      {
+        "transaction_id": 0,
```
You are not documenting that you are using the network-events data format. Please modify the code such that we are not using the `transaction_id` for a different purpose than the one for which it is usually used.

I recommend you introduce your own `x_something` code for this experiment to keep the number that tells you the state of the handshake in which you are (or just drop the number?).

I think we should also clarify what happens in case of error. Does the failure in, say, stage 3 causes a `failure` to be written in the corresponding `network_events` stage?

ainghazal replied:
https://github.com/ooni/spec/pull/255#discussion_r1018902876 I will revert to use `x_operation_id`. right now there is no failure inside the `network_event` (the failure is stored as the experiment `failure`, but it looks like this is the right thing to do).

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+      }
+    ],
+    "requests": [
+    // ...
```
Set the data type in the ``` using JavaScript. This will make comments render without errors.

bassosimone replied:
Maybe you can show a single request for clarity?

ainghazal replied:
there's a request below, in "example output sample". do you think it's still needed here?

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+    "requests": [
+    // ...
+    ],
+    "minivpn_version": "(devel)",
```
I am not super sold about emitting a non version number as the version

ainghazal replied:
this is annoying yes. it's a side-effect of using `go.work` locally for overrides. I'll investigate if I can fetch the original version somehow.

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+- `provider`, `vpn_protocol`, `transport`, `remote` and `obfuscation` are
+  parameters extracted by the probe from the input `URI` (and as such, they
+  should match). However, the authoritative source is the entry in the
+  test-keys.
```
Please add an empty line between each entry

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+- `tcp_connect`: info about the TCP handshake (only when `transport` == `tcp`). Conforms to `df-005-tcpconnect`.
+- `icmp_pings` is an array containing the result for a series of `icmp` pings through the tunnel:
+    - `target` (`string`): the IP the ICMP trains were targeting,
+    - `sequence` is an aray with the ping responses:
```
```suggestion
    - `sequence` is an array with the ping responses:
```

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+- `obfuscation` (`string`): type of obfuscation in use. Currently known
+  obfuscation schemes include `none` and `obfs4`,
+- `bootstrap_time` (`float`) is the time, in seconds, to bootstrap a VPN connection,
+- `success` (`bool`) whether all the stages in the experiment were successful,
```
can you consistently use `.` instead of mixing `.` and `,` to signal the end of an item?

bassosimone commented on nettests/ts-039-openvpn.md:
```diff
@@ -0,0 +1,688 @@
+# Packet capture considerations
+
+This test does not capture packets by default.
+
```
```suggestion
```

bassosimone reviewed (commented):
Thanks for writing this spec! 🙌 

I have provided comments and suggestions!

ainghazal commented:
closing now that we merged #293
